### PR TITLE
st_unaryunion implementation for ESRI/JTS

### PR DESCRIPTION
**Pull request by tdikland**
Implementation of the `st_unaryunion` spatial function.

Partially implements #209

### CODE REVIEW

**Review by milos-colic — APPROVED** [review]
@tdikland this is a great first PR.
the only thing that it took me a bit to understand was geom.union(geom) in ESRI.
Maybe worth an inline comment saying this intended way of performing a unary union in ESRI (given we are calling a binary function with same geom on both sides).
The rest is super clear.

File: src/main/scala/com/databricks/labs/mosaic/core/geometry/MosaicGeometryESRI.scala [hunk 1]
```diff
@@ -127,6 +127,8 @@ abstract class MosaicGeometryESRI(geom: OGCGeometry) extends MosaicGeometry {
 
     override def convexHull: MosaicGeometryESRI = MosaicGeometryESRI(geom.convexHull())
 
+    override def unaryUnion: MosaicGeometry = MosaicGeometryESRI(geom.union(geom))
```
Comment (milos-colic) [inline src/main/scala/com/databricks/labs/mosaic/core/geometry/MosaicGeometryESRI.scala:130]: @tdikland I would add here a one liner to clarify why geom.union(geom) in ESRI is equivalent to geom.union() in JTS (I would argue JTS way is more intuitive).

**Review by milos-colic — APPROVED** [review]

**Review by edurdevic — CHANGES_REQUESTED** [review]
Looks mostly good to me!
Please take a look at the unit tests and assert that result matches the expected value for the codegen tests

File: src/test/scala/com/databricks/labs/mosaic/expressions/geometry/ST_UnaryUnionTest.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        ) _
+
+    private val codegenOnly =
+        withSQLConf(
+          SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "false",
+          SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true",
+          SQLConf.CODEGEN_FACTORY_MODE.key -> CodegenObjectFactoryMode.CODEGEN_ONLY.toString
+        ) _
+
+    test("Testing stUnaryUnion (H3, JTS) NO_CODEGEN") { noCodegen { unaryUnionBehavior(H3IndexSystem, JTS) } }
```
[the remark this answers is not in the document]
Comment (milos-colic) [reply src/test/scala/com/databricks/labs/mosaic/expressions/geometry/ST_UnaryUnionTest.scala:24]: We always do, to ensure there is no edge case.
Logic wise it is redundant but binary wise it may not be.
We are mixing Index Systems and Geometry Engines at run time.
Better test for each expression.

File: src/test/scala/com/databricks/labs/mosaic/expressions/geometry/ST_UnaryUnionBehaviours.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        val sc = spark
+        import mc.functions._
+        import sc.implicits._
+        mc.register(spark)
+
+        val multiPolygon = List("MULTIPOLYGON (((10 10, 20 10, 20 20, 10 20, 10 10)), ((15 15, 25 15, 25 25, 15 25, 15 15)))").toDF("multipolygon")
+        val expected = List("POLYGON ((20 15, 20 10, 10 10, 10 20, 15 20, 15 25, 25 25, 25 15, 20 15))").toDF("polygon")
+
+        val result = multiPolygon
+          .crossJoin(expected)
```
[the remark this answers is not in the document]
Comment (tdikland) [reply src/test/scala/com/databricks/labs/mosaic/expressions/geometry/ST_UnaryUnionBehaviours.scala:48]: This was an error during copy-pasting. These testcases should only check compilation of the generated code. I removed the expectation and the crossJoin consequently.

File: src/main/scala/com/databricks/labs/mosaic/core/geometry/MosaicGeometryESRI.scala (hunk 1 above)
Comment (tdikland) [reply src/main/scala/com/databricks/labs/mosaic/core/geometry/MosaicGeometryESRI.scala:130]: Completely agree. Added some inline explanations.

**Comment by tdikland** [comment]
Thanks @edurdevic and @milos-colic for your very helpful feedback. I have made the requested changes, and left the testing to check all available grid index systems.

**Review by edurdevic — APPROVED** [review]
Looks good! Well done @tdikland!
